Improve 'build' and 'osc build' experience

e4t opened the pull request:
This set contains two things:

1. Improvement of the 'build' and 'osc build' experience: together with changes to github.com/openSUSE/osc and if supported for the underlying build system, allow for 'partial' builds - like rpmbuild --short-circuit -b<build_stage>, but better: for multi-stage builds, it allows to execute:
    - a single build stage, 
    - a build up to a specified stage or 
    - a build starting from a specified stage (provided all previous stages have already been performed).
No more waiting for endless %builds just to test a change to an %files stage, no more need to fiddle with rpmbuild directly in the chroot (and forgetting to copy back the changes to the spec file later).
(Note: this is presently available for rpmbuild only, the design is sufficiently generic that it should be implementable to every other build system - provided this supports multiple build stages and running them individually.
2. Additions to the documentation to document the changes in (1.) as well as some other missing bits and pieces.

mlschroe commented:
Hmm, can't say I like the user interface with the new 'mode' option. And I also dislike the old '--stage' option, as you always need to prefix the stage with `-b`.
I would rather have extend --stage so that you it offers to specify a range, like:
```
--stage=i    only do the install stage
--stage=-i   do all stages up to install
--stage=i-   do stages starting from install
```
We can also use some different char for the range, so that it does not look that much like an option.

Of course we need to stay backward compatible, so that --stage=-bX still works.

mlschroe commented on build-recipe-spec:
```diff
@@ -21,13 +21,75 @@
+               -bp) test $BUILD_RPM_BUILD_MODE = post && BUILD_RPM_BUILD_STAGE=-ba
+                    BUILD_RPM_BUILD_MODE=pre ;;
+               -bc) pre_d="BUILD"; post_s="-bc -bi -bb -bs" ;;
+               -bi) pre_d="BUILD"; post_s="-bi -bs -bb" ;;
```
-bs should always come atfer -bb

e4t replied:
Ok. will fix.

mlschroe commented on build-recipe-spec:
```diff
@@ -126,91 +187,97 @@ recipe_build_spec() {
-    check_exit
-    chroot $BUILD_ROOT su -c /.build.command - $BUILD_USER < /dev/null
-    st=$?
+	test "$st" = 0 -o "$st" = 11 || break
```
Why do you also test for `11` here?

e4t replied:
Right, on 11 the caller should probably handle this. I will fix this.

e4t commented:
I'll change the semantics of --stage according to Michael's suggestion (of course I will keep the old semantics intact).

e4t commented:
I've updated the patch. Instead of a `--mode=...` option, I've introduced 3 new forms of the `--stage` option:
- `--stage=..STAGE` to build up to STAGE,
- `--stage=STAGE` to build only STAGE,
- `--stage=STAGE..` to build everything starting from STAGE.

All other issues were also addressed.

mlschroe commented:
OK, I somewhat rewrote the code and manually merged it. I did change the stage selection, because I think that specifying a single letter stage should to the expected thing, i.e. just pass -bSTAGE to rpm and build the stages up to and including the specified stage.

So I needed to come up with ways to select "just this stage" and "this stage and the stages after it". I went with a trailing '+' to specify the latter and a trailing '=' to specify single stage builds. I'm not super happy about the '=', though. And I'm also wondering if we should have a 'everything after this stage excluding the specified stage' flag, because i can imaging that this is more useful than what we have. Suggestions welcome.

Anyway, thank you very much for opening this pull request!

e4t commented:
I personally have the greatest need for this for the the %install stage. The install stage sets the 'stage' for everything that comes later. If a file is moved during %install my %files sections may no longer work. Technically, these things usually get fixed by adjustments to the %files section, however, running the this section without doing a prior install usually fails. However, %install takes only a fraction of the time of %build usually, so this is not doing exactly what I need. I'm fine with it.

Thanks for looking into it!
